File: model/features.py
"""
Board state → tensor encoding for the neural network.

Converts a Board object into a multi-channel tensor that the ResNet can process.
Each channel is a 2D binary (or scalar) plane of size board_size x board_size.

Feature planes:
  0: current player's stones (1 where current player has a stone)
  1: opponent's stones
  2: empty points
  3: last move (one-hot, all zeros if no moves yet or last was pass)
  4: color to play (all 1s if black, all 0s if white)
  5: liberties == 1 (stones in atari)
  6: liberties == 2
  7: liberties >= 3

Total: 8 channels
"""

import numpy as np
import torch

from go_engine.board import Board, BLACK, WHITE, EMPTY, OPPONENT

# Try to import C++ feature extraction
try:
    import sys, os
    sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
    import alphago_core as _ac
    _HAS_CPP_FEATURES = True
except ImportError:
    _HAS_CPP_FEATURES = False


NUM_FEATURES = 8


def board_to_features(board: Board, color_to_play: int) -> np.ndarray:
    """
    Convert a board state to a feature tensor.

    Args:
        board: Current board state
        color_to_play: BLACK or WHITE (whose turn it is)

    Returns:
        numpy array of shape (NUM_FEATURES, board_size, board_size), dtype float32
    """
    size = board.size
    opp = OPPONENT[color_to_play]
    features = np.zeros((NUM_FEATURES, size, size), dtype=np.float32)

    # Plane 0: current player's stones
    features[0] = (board.grid == color_to_play).astype(np.float32)

    # Plane 1: opponent's stones
    features[1] = (board.grid == opp).astype(np.float32)

    # Plane 2: empty points
    features[2] = (board.grid == EMPTY).astype(np.float32)

    # Plane 3: last move (one-hot)
    if board.move_history:
        _, last_move = board.move_history[-1]
        if last_move is not None:
            r, c = last_move
            features[3, r, c] = 1.0

    # Plane 4: color to play (all 1s for black, all 0s for white)
    if color_to_play == BLACK:
        features[4] = 1.0

    # Planes 5-7: liberty features for ALL stones on the board
    # We compute liberties per group and mark each stone in the group
    visited = np.zeros((size, size), dtype=bool)
    for r in range(size):
        for c in range(size):
            if board.grid[r, c] == EMPTY or visited[r, c]:
                continue
            stones, liberties = board._group(r, c)
            lib_count = len(liberties)
            for sr, sc in stones:
                visited[sr, sc] = True
                if lib_count == 1:
                    features[5, sr, sc] = 1.0
                elif lib_count == 2:
                    features[6, sr, sc] = 1.0
                elif lib_count >= 3:
                    features[7, sr, sc] = 1.0

    return features


def board_to_features_cpp(board, color_to_play: int) -> np.ndarray:
    """
    C++ accelerated feature extraction.
    Works with both CppBoard (has ._board) and CBoard directly.
    """
    # Get the underlying C++ board
    cboard = getattr(board, '_board', None)
    if cboard is None and hasattr(board, 'at'):
        # It's already a CBoard
        cboard = board

    if cboard is None:
        # Fall back to Python
        return board_to_features(board, color_to_play)

    # Get last move for plane 3
    last_r, last_c = -1, -1
    if hasattr(board, 'move_history') and board.move_history:
        _, last_move = board.move_history[-1]
        if last_move is not None:
            last_r, last_c = last_move

    return _ac.board_to_features_cpp(cboard, color_to_play, last_r, last_c)


def board_to_tensor(board, color_to_play: int) -> torch.Tensor:
    """
    Convert board state to a PyTorch tensor ready for the network.

    Uses C++ feature extraction when available for ~10x speedup.

    Returns:
        Tensor of shape (1, NUM_FEATURES, board_size, board_size)
        (batch dimension included)
    """
    if _HAS_CPP_FEATURES and hasattr(board, '_board'):
        features = board_to_features_cpp(board, color_to_play)
    else:
        features = board_to_features(board, color_to_play)
    return torch.from_numpy(features).unsqueeze(0)


def apply_symmetry(features: np.ndarray, policy: np.ndarray, sym_index: int, board_size: int) -> tuple[np.ndarray, np.ndarray]:
    """
    Apply one of 8 board symmetries (4 rotations x 2 reflections) to both
    feature planes and the policy vector.

    This gives us 8x training data for free since Go is symmetric.

    Args:
        features: shape (C, size, size)
        policy: shape (size*size + 1,) — board moves + pass
        sym_index: 0-7 (0 = identity)
        board_size: size of the board

    Returns:
        (transformed_features, transformed_policy)
    """
    # Separate pass probability from board probabilities
    board_policy = policy[:board_size * board_size].reshape(board_size, board_size)
    pass_prob = policy[-1]

    # Apply symmetry to feature planes and policy board
    f = features.copy()
    p = board_policy.copy()

    if sym_index >= 4:
        # Reflect horizontally
        f = f[:, :, ::-1].copy()
        p = p[:, ::-1].copy()

    rotations = sym_index % 4
    if rotations > 0:
        f = np.rot90(f, k=rotations, axes=(1, 2)).copy()
        p = np.rot90(p, k=rotations).copy()

    # Reconstruct policy vector
    transformed_policy = np.append(p.flatten(), pass_prob)

    return f, transformed_policy

File: src/features.cpp
#include "go_features.h"
#include <cstring>

namespace alphago {

BoardFeatures board_to_features(
    const Board& board,
    uint8_t color_to_play,
    Move last_move
) {
    BoardFeatures f;
    f.board_size = board.size();
    std::memset(f.data, 0, sizeof(f.data));

    int size = board.size();
    uint8_t opp = opponent(color_to_play);

    // Planes 0-2: stones and empty
    for (int r = 0; r < size; r++) {
        for (int c = 0; c < size; c++) {
            uint8_t cell = board.at(r, c);
            if (cell == color_to_play) f.data[0][r][c] = 1.0f;
            else if (cell == opp)      f.data[1][r][c] = 1.0f;
            else                       f.data[2][r][c] = 1.0f;
        }
    }

    // Plane 3: last move
    if (!last_move.is_pass()) {
        f.data[3][last_move.row][last_move.col] = 1.0f;
    }

    // Plane 4: color to play
    if (color_to_play == BLACK) {
        for (int r = 0; r < size; r++)
            for (int c = 0; c < size; c++)
                f.data[4][r][c] = 1.0f;
    }

    // Planes 5-7: liberty features
    // Use the board's public interface to count liberties
    bool visited[MAX_BOARD_SIZE][MAX_BOARD_SIZE] = {};
    for (int r = 0; r < size; r++) {
        for (int c = 0; c < size; c++) {
            if (board.at(r, c) == EMPTY || visited[r][c]) continue;

            int lib_count = board.count_liberties(r, c);

            // Mark all stones in this group via simple flood fill
            int plane = (lib_count == 1) ? 5 : (lib_count == 2) ? 6 : 7;

            // BFS to find connected stones of same color
            uint8_t color = board.at(r, c);
            std::vector<std::pair<int,int>> stack;
            stack.push_back({r, c});
            visited[r][c] = true;

            while (!stack.empty()) {
                auto [cr, cc] = stack.back();
                stack.pop_back();
                f.data[plane][cr][cc] = 1.0f;

                // Check 4 neighbors
                int dr[] = {-1, 1, 0, 0};
                int dc[] = {0, 0, -1, 1};
                for (int d = 0; d < 4; d++) {
                    int nr = cr + dr[d], nc = cc + dc[d];
                    if (nr >= 0 && nr < size && nc >= 0 && nc < size
                        && !visited[nr][nc] && board.at(nr, nc) == color) {
                        visited[nr][nc] = true;
                        stack.push_back({nr, nc});
                    }
                }
            }
        }
    }

    return f;
}

void apply_symmetry(
    const float* features_in,
    const float* policy_in,
    float* features_out,
    float* policy_out,
    int channels,
    int board_size,
    int sym_index
) {
    int N = board_size;
    int board_cells = N * N;

    auto transform_plane = [&](const float* in, float* out) {
        for (int r = 0; r < N; r++) {
            for (int c = 0; c < N; c++) {
                int sr = r, sc = c;

                if (sym_index >= 4) {
                    sc = N - 1 - sc;
                }

                int rotations = sym_index % 4;
                for (int rot = 0; rot < rotations; rot++) {
                    int tmp = sr;
                    sr = sc;
                    sc = N - 1 - tmp;
                }

                out[sr * N + sc] = in[r * N + c];
            }
        }
    };

    for (int ch = 0; ch < channels; ch++) {
        transform_plane(
            features_in + ch * board_cells,
            features_out + ch * board_cells
        );
    }

    float board_policy_in[MAX_BOARD_SIZE * MAX_BOARD_SIZE];
    float board_policy_out[MAX_BOARD_SIZE * MAX_BOARD_SIZE];

    std::memcpy(board_policy_in, policy_in, board_cells * sizeof(float));
    transform_plane(board_policy_in, board_policy_out);

    std::memcpy(policy_out, board_policy_out, board_cells * sizeof(float));
    policy_out[board_cells] = policy_in[board_cells];
}

} // namespace alphago
